netebpfext/net_ebpf_ext.c

```diff
@@ -998,4 +1127,37 @@ net_ebpf_ext_remove_client_context(
     }
 
     ExReleaseSpinLockExclusive(&filter_context->lock, old_irql);
+}
+
+void
+net_ebpf_ext_add_provider_context_to_cleanup_list(_Inout_ net_ebpf_extension_hook_provider_t* provider_context)
+{
+    KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock);
+    InsertTailList(
+        &_net_ebpf_ext_wfp_cleanup_state.provider_context_cleanup_list, &provider_context->cleanup_list_entry);
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock, old_irql);
+}
+
+void
+net_ebpf_ext_add_filter_context_to_cleanup_list(_Inout_ net_ebpf_extension_wfp_filter_context_t* filter_context)
+{
+    KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock);
+    InsertTailList(&_net_ebpf_ext_wfp_cleanup_state.filter_cleanup_list, &filter_context->link);
+    filter_context->in_cleanup_list = TRUE;
+    ExReleaseSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock, old_irql);
+}
+
+void
+net_ebpf_ext_remove_filter_context_from_cleanup_list(_Inout_ net_ebpf_extension_wfp_filter_context_t* filter_context)
+{
+    if (filter_context->in_cleanup_list) {
+        KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_wfp_cleanup_state.lock);
+        RemoveEntryList(&filter_context->link);
+        filter_context->in_cleanup_list = FALSE;
```

Comment:
Just an idea, but you maybe be able to eliminate this `in_cleanup_list` variable and avoid dangling pointers to the `filter_cleanup_list` by re-initializing the `link` to an empty list, and treating the empty list as a sentinel value.

(`RemoveEntryList` doesn't modify the provided list entry's pointer fields, only the pointers in the rest of the list.)